src/backend/cuda/kernel/join.hpp

```diff
@@ -73,6 +73,15 @@ namespace cuda
                         blocksPerMatY * X.dims[3],
                         1);
 
+            int maxBlocksY = cuda::getDeviceProp(cuda::getActiveDeviceId()).maxGridSize[1];
+            // Increase the work per block until the condition is satisfied
+            // Increase threads only upto 512
+            while(blocks.y > maxBlocksY) {
```

Comment:
You were dividing old `blocksPerMatY` by  `2`, this one divides the `maxBlockSize` by `X.dims[3]` which ensures that it falls within the given size.